docs/language-reference/04-types.md

```diff
@@ -0,0 +1,323 @@
+On all platforms, the `bool` type must be _naturally aligned_ (its alignment is its size).
+
+### Integer Types
+
+The following integer types are defines:
+
+| Name          | Description |
+|---------------|-------------|
+| `int8_t`      | 8-bit signed integer |
+| `int16_t`     | 16-bit signed integer |
+| `int`         | 32-bit signed integer |
+| `int64_t`     | 64-bit signed integer |
+| `uint8_t`     | 8-bit unsigned integer |
+| `uint16_t`    | 16-bit unsigned integer |
+| `uint`        | 32-bit unsigned integer |
+| `uint64_t`    | 64-bit unsigned integer |
+
+All signed integers used two's complement representation.
+All arithmetic operations on integers (both signed and unsigned) wrap on overflow/underflow.
+
+All target platforms must support the `int` and `uint` types.
+Specific target platforms may not support the other integer types.
+
+All integer types are stored in memory with their natural size and alignment on all targes that support them.
+
+### Floating-Point Types
+
+The following floating-point type are defined:
+
+| Name          | Description                   |
+|---------------|-------------------------------|
+| `half`        | 16-bit floating-point number (1 sign bit, 5 exponent bits, 10 fraction bits) |
+| `float`       | 32-bit floating-point number (1 sign bit, 8 exponent bits, 23 fraction bits) |
+| `double`      | 64-bit floating-point number (1 sign bit, 11 exponent bits, 52 fraction bits) |
+
+All floating-point types are laid out in memory using the matching IEEE 754 standard format (`binary16`, `binary32`, `binary64`).
+Target platforms may define their own rules for rounding, precision, denormals, infinities, and not-a-number values.
+
+All target platforms must support the `float` type.
+Specific targets may not support the other floating-point types.
+
+All floating-point types are stored in memory with their natural size and alignment on all targets that support them.
+
+Vector Types
+------------
+
+A vector type is written as `vector<T, N>` and represents an `N`-element vector with elements of type `T`.
+The _element type_ `T` must be one of the built-in scalar types, and the _element count_ `N` must be a specialization-time constant integer.
+The element count must be between 2 and 4, inclusive.
+
+A vector type allows subscripting of its elements like an array, but also supports component-wise arithmetic on its elements.
+
+A vector type is laid out in memory as `N` contiguous values of type `T` with no padding.
+The alignment of a vector type may vary by target platforms.
+The alignment of `vector<T,N>` will be at least the alignment of `T` and may be at most `N` times the alignment of `T`.
+
+As a convenience, Slang defines built-in type aliases for vectors of the built-in scalar types.
+E.g., declarations equivalent to the following are provided by the Slang standard library:
+
+```hlsl
+typealias float4 = vector<float, 4>;
+typealias int8_t3 = vector<int8_t, 3>;
```

Comment:
Don't complain to me; this is straight up what is in HLSL as implemented by dxc.

There is an important point here, though, in that there is a conflict between two different conventions for what suffix digits on a type name mean:

* The shading language convention is that the digits tell us the number of elements in a vector (or rows x columns in a matrix)

* The dominant convention for every other systems-y language on the planet is that the digits tells us the number of bits in a scalar type

The result of these dualing conventions is that you have things like an OpenCL `float32` being a 32-element vector of `float` while a Swift `Float32` is a 32-bit floating-point scalar.

I am not aware of any precedent for a language that has been able to marry the conventions successfully, and the solution of copy-pasting the `_t` suffix from C over to shading languages sure feels like a "clever" cheat rather than something pleasant.

Cg and CUDA seem to handle the problem by defining old-fashioned C-style names for all the types of different sizes (e.g, `char`, `short`, `int`, `long`) so that vector names don't feel clunky. That kind of works, but it doesn't seem to reflect the trajectory that general-purpose languages have taken of being more explicit about the size of fundamental types.

The GLSL convention actually has some promise here in that it actually *doesn't* just use a numeric suffix and instead uses the word `vec` or `mat` before the suffix, for all types. GLSL instead uses a per-type *prefix* before the `vec` or `mat`, with the shorter prefixes being reserved for more common types.

I could get behind a convention of defining stuff like `Int8Vec4`, `Float16Mat2x2`, etc. as a convention for these types. It is still a mouthful, but far better than `int8_t4` or `float16_t2x2`.